cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,236 @@
+          "else return nil end\n"
+      ));
+
+      // If the agent isn't present in either the WAITING or WORKING sets, it's safe to add. If it's present in either,
+      // it's being worked on or was recently run, so leave it be.
+      // KEYS[1] and KEYS[2] are checked for inclusion. If the agent is in neither ARGV[1] is added to KEYS[1] with score
+      // ARGV[2].
+      scriptShas.put(ADD_AGENT_SCRIPT, jedis.scriptLoad(
+        "if redis.call('zrank', KEYS[1], ARGV[1]) ~= nil then\n" +
+        "  if redis.call('zrank', KEYS[2], ARGV[1]) ~= nil then\n" +
+        "    return redis.call('zadd', KEYS[1], ARGV[2], ARGV[1])\n" +
+        "  else return nil end\n" +
+        "else return nil end\n"));
+    }
+  }
+
+  @Override
+  public void schedule(Agent agent, AgentExecution agentExecution, ExecutionInstrumentation executionInstrumentation) {
+    if (agent instanceof AgentSchedulerAware) {
+      ((AgentSchedulerAware)agent).setAgentScheduler(this);
+    }
+    agents.put(agent.getAgentType(), new AgentWorker(agent, agentExecution, executionInstrumentation, this));
+    try (Jedis jedis = jedisSource.getJedis()) {
+      jedis.evalsha(scriptShas.get(ADD_AGENT_SCRIPT), 2, WAITING_SET, WORKING_SET, agent.getAgentType(), score(NOW));
+    }
+  }
+
+  @Override
+  public void run() {
+    if (!nodeStatusProvider.isNodeEnabled()) {
+      return;
+    }
+    try {
+      saturatePool();
+    } catch (Throwable t) {
+      log.error("Failed to run caching agents", t);
+    } finally {
+      runCount++;
+    }
+  }
+
+  private static String score(long offset) {
+    return String.format("%d", (TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()) + offset));
+  }
+
+  private boolean acquireAgent(Agent agent) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      return jedis.evalsha(scriptShas.get(SWAP_SET_SCRIPT),
+          Arrays.asList(WAITING_SET, WORKING_SET),
+          Arrays.asList(agent.getAgentType(), score(intervalProvider.getInterval(agent).getTimeout()))) != null;
+    }
+  }
+
+  private boolean releaseAgent(Agent agent) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      if (jedis.evalsha(scriptShas.get(SWAP_SET_SCRIPT),
+          Arrays.asList(WORKING_SET, WAITING_SET),
+          Arrays.asList(agent.getAgentType(), score(intervalProvider.getInterval(agent).getTimeout()))) != null) {
+        runningAgents.release();
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  private void saturatePool() {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      // Periodically repopulate redis with everything this agent is aware of (in case redis was flushed or crashed).
+      if (runCount % REDIS_REFRESH_PERIOD == 0) {
+        for (String type : agents.keySet()) {
+          jedis.evalsha(scriptShas.get(ADD_AGENT_SCRIPT), 2, WAITING_SET, WORKING_SET, type, score(NOW));
+        }
+
+        storeScripts();
```

Comment:
maybe do this before the for loop - on the chance it was needed it would mean the script gets repopulated before the evalsha call in the loop